# VCFtoSummStats

A tool for generating a "tidy", plain-text file of allele frequencies from a VCF.

## Building the program

**Building the program requires the g++ compiler.**

You can build (compile) the program from the UNIX/Linux-style command line by
1. cloning this repo
2. cd'ing to it on your command line
3. typing "make" at the command line and pressing return/enter


## Running the program

Use of the helper wrapper script is REQUIRED (at least once), because it creates the input parameters (arguments) to the program, as well as a necessary simple file with some population metadata.  

The wrapper script is invoked in the UNIX/Linux-like terminal with a command like this:

```bash WrapperForVCFtoSummStats.sh path/to/VCFfile.vcf path/to/samplesAndPopulations.txt```

where `VCFfile.vcf` is your VCF file, and `samplesAndPopulations.txt` is a two-column, 
whitespace delimited, plain-text datafile, in which the first column is the sample IDs 
(corresponding to columns ID's of sample data in the VCF file) and the second column 
is a population designation for each sample (more details below).
The "path/to/" should be the same for both files.
(See below for an additional example.)

## Assumptions about VCF file format
The program assumes that the VCF file supplied to the program follows the VCF v4.3 format guidelines as found at [http://samtools.github.io/hts-specs/VCFv4.3.pdf](http://samtools.github.io/hts-specs/VCFv4.3.pdf), accessed 5/31/19.

The wrapper script does do some minimal checking to make sure that the columns/fields are as expected.


## Requirements of file with information on samples and populations

The second argument to the wrapper script (as shown in the example command above) is the 
name of a file containing information on the samples and the population or taxonomic
group to which each sample belongs.  The sample IDs must be in the first column, and
the population/taxonomic designation must be in the second column. 
The columns must be separated by whitespace. 
The file should **NOT have a header**, but if it does, add a third argument, "Y", to the call 
to the script (i.e., call the script with
`bash WrapperForVCFtoSummStats.sh path/to/VCFfile.vcf path/to/samplesAndPopulations.txt Y`).
The "Y" tells the script that, yes, there is a header.  The default is "n" (for no).


## Example data files provided here
An example VCF and population designation file are provided in the `ExampleDataFiles/` directory here.  The VCF is a subset of a much larger file from the data archive of Schilling et al. 2018 (_Genes_ 2018, 9(6), 274).  
The original publication is freely available at: [https://doi.org/10.3390/genes9060274](https://doi.org/10.3390/genes9060274)
And the data archive is at: [http://bit.ly/2s6jeIf](http://bit.ly/2s6jeIf)

After cloning this repo, you should be able to run the program (with this directory 
as your current working directory) by using the command syntax given above.  
An example that you can use to test it quickly on your system is the following:

```bash WrapperForVCFtoSummStats.sh ExampleDataFiles/Small_hmel2.5.30f4.vcf ExampleDataFiles/popFileHmel.txt```

That command will cause the wrapper script to run, 
which will call the program with appropriate arguments.
The main results of that command, assuming it runs successfully on your system, 
will be output to a file: `ExampleDataFiles/Small_hmel2.5.30f4.vcf_Unfiltered_Summary.tsv`.
Some other accessory files will also be generated by the script and program.



